### PR TITLE
[WIP] Add basic support for SourceForge

### PR DESCRIPTION
**Pull request by swissspidy**
Because we can 🤷‍♀️

SourceForge supports Git, Mercurial, and Subversion. All at once or none at all. That means people will still need to hook into `traduttore.loader` to choose the correct source code loader.

### CODE REVIEW

File: inc/Repository/SourceForge.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+	/**
+	 * Returns the repository type.
+	 *
+	 * @since 3.0.0
+	 *
+	 * @return string Repository type.
+	 */
+	public function get_type() : string {
+		return Repository::TYPE_GITLAB;
```
Comment (grappler) [inline inc/Repository/SourceForge.php]:
```suggestion
		return Repository::TYPE_SOURCEFORGE;
```

File: inc/Repository/SourceForge.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	 * @return string Repository host name.
+	 */
+	public function get_host(): string {
+		$host = parent::get_host();
+
+		if ( $host ) {
+			return $host;
+		}
+
+		return 'gitlab.com';
```
Comment (grappler) [inline inc/Repository/SourceForge.php]:
```suggestion
		return 'sourceforge.net';
```

File: inc/Repository/SourceForge.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ * SourceForge repository class.
+ *
+ * @since 3.0.0
+ *
+ * @package Required\Traduttore
+ */
+
+namespace Required\Traduttore\Repository;
+
+use Required\Traduttore\Repository;
```
Comment (grappler) [inline inc/Repository/SourceForge.php]: Is this needed if it is in the same namespace?

Comment (swissspidy) [reply inc/Repository/SourceForge.php]: That class is not in the same namespace 🧐 The class just has the same name as the namespace.

**Comment by swissspidy** [comment]
> Is there a reason why this is not failing the tests?

Which ones should fail in your opinion? I didn't add any loader tests for this one (because of the manual involvement) and existing ones shouldn't be affected. Plus, existing loader tests are currently skipped anyway.

**Comment by grappler** [comment]
>  I didn't add any loader tests for this one (because of the manual involvement)

Ok, that makes sense now.

File: docs/hooks.md [hunk 4]
```diff
@@ -152,3 +152,42 @@ Filters the delay for scheduled project updates.
...
 * `$delay`: Delay in minutes. Default is 3 minutes.
 * `$project`: The current project.
+
+----
+
+### `traduttore.loader`
+
+**Since:** 3.0.0
+
+Filters the loader instance for a given repository and project.
```
Comment (grappler) [inline docs/hooks.md]: Initially I was not sure what the loader instance ist or what the loader class does. Maybe add something like that it is to do with the VCS that is used for the repository.

File: docs/hooks.md [hunk 5]
```diff
@@ -152,3 +152,42 @@ Filters the delay for scheduled project updates.
...
+* `$handler`: Webhook handler instance.
+* `$request`: The current request object.
+
+----
+
+### `traduttore.repository`
+
+**Since:** 3.0.0
+
+Filters the determined repository instance for a given project.
```
Comment (grappler) [inline docs/hooks.md]:
```suggestion
Filters the determined repository instance for a given project. Can be used to set a custom handler for self-managed repositories.
```

File: docs/hooks.md [hunk 6]
```diff
@@ -152,3 +152,42 @@ Filters the delay for scheduled project updates.
...
+
+### `traduttore.webhook_handler`
+
+**Since:** 3.0.0
+
+Filters the determined incoming webhook handler.
+
+**Parameters:**
+
+* `$handler`: Webhook handler instance.
```
Comment (grappler) [inline docs/hooks.md]:
```suggestion
* `$handler`: WebhookHandler instance or null.
```

File: docs/hooks.md [hunk 7]
```diff
@@ -152,3 +152,42 @@ Filters the delay for scheduled project updates.
...
+* `$request`: The current request object.
```
Comment (grappler) [inline docs/hooks.md]:
```suggestion
* `$request`: The current WP_REST_Request object.
```

**Comment by swissspidy** [comment]
> Do you have a repo that we can use to test that it is working?

I've set up https://sourceforge.net/projects/pascals-test-project/ but with no webhooks configured yet. It's on my todo list :-)

I also have two GitLab repositories that I still want to properly test for #61.

File: docs/hooks.md (hunk 5 above)
Comment (swissspidy) [reply docs/hooks.md]: Thanks. Will add this on a new line, like in the inline docs.

File: inc/WebhookHandler/SourceForge.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	 *
+	 * @return WP_Error|WP_REST_Response REST response on success, error object on failure.
+	 */
+	public function callback() {
+		$params = $this->request->get_params();
+
+		// We only care about the default branch (we have to assume it's master) but don't want to send an error still.
+		if ( 'refs/heads/master' !== $params['ref'] ) {
+			return new WP_REST_Response( [ 'result' => 'Not the default branch' ] );
+		}
```
Comment (swissspidy) [inline inc/WebhookHandler/SourceForge.php:62]: Might be worth making the default branch name filterable / configurable. Might also be relevant for others.

**Pull request closed without merging**